lib/claim/token-holder-revenue-fund-contracts-ensemble.js

```diff
@@ -0,0 +1,359 @@
+'use strict';
+
+const {utils: {bigNumberify}} = require('ethers');
+const TokenHolderRevenueFundContract = require('./token-holder-revenue-fund-contract');
+const NestedError = require('../nested-error');
+
+const _ensemble = new WeakMap();
+const _spannedByCurrency = new WeakMap();
+const _firstAccrualOffset = new WeakMap();
+
+const currencyToKey = (currency) => `${currency.ct.toString()}-${currency.id}`;
+
+const minBigNumber = function (n1, n2) {
+    return n1.lt(n2) ? n1 : n2;
+};
+
+/**
+ * @class TokenHolderRevenueFundContractsEnsemble
+ * An ensemble of instances of TokenHolderRevenueFundContract
+ * @alias module:nahmii-sdk
+ */
+class TokenHolderRevenueFundContractsEnsemble {
+    constructor(walletOrProvider, abstractionNames = ['TokenHolderRevenueFund'], firstAccrualOffset = 0) {
+        _ensemble.set(this, abstractionNames.map(n => ({contract: new TokenHolderRevenueFundContract(walletOrProvider, n)})));
+        _spannedByCurrency.set(this, new Map());
+        _firstAccrualOffset.set(this, firstAccrualOffset > 0 ? firstAccrualOffset : 0);
+    }
+
+    /**
+     * Returns the first accrual's offset index used in the (temporal) span
+     * @returns {number} - First accrual's offset
+     */
+    get firstAccrualOffset() {
+        return _firstAccrualOffset.get(this);
+    }
+
+    /**
+     * Return status of whether contract instance span has been executed
+     * for the given currency
+     * @param {Currency} currency - The currency
+     * @returns {boolean} True if span has been executed, else false
+     */
+    isSpanned(currency) {
+        const spannedByCurrency = _spannedByCurrency.get(this);
+        return spannedByCurrency.get(currencyToKey(currency)) || false;
+    }
+
+    /**
+     * Span (temporally) in accruals and block numbers for the given currency. I.e. infer the
+     * accrual and block number boundaries of the given currency for each contract
+     * instance in the ensemble
+     * @param {Currency} currency - The currency
+     */
+    async span(currency) {
+        const ensemble = _ensemble.get(this);
+        let globalAccrualOffset = _firstAccrualOffset.get(this);
+        const currencyKey = currencyToKey(currency);
+
+        for (const configuration of ensemble) {
+            const closedAccrualsCount = (await configuration.contract.closedAccrualsCount(
+                currency.ct.toString(), currency.id
+            )).toNumber();
+
+            const span = {};
+            span.startAccrual = globalAccrualOffset || 0;
```

Comment:
Pedantic comment:
I find it useful to constrain myself to name ranges in one of the following forms
    Ruby index range style `[first..last]` where `count === last - first + 1`, or
    C++ index range style `[begin..end>` where `count === end - begin`
    Temporal style `[start..stop>` where `duration === stop - start`

Reply:
I know you are pedantic in this respect.😄 Nonetheless, I find your comments relevant. The usage of `start...` or `end...` parameters is more in line with _first_ and _last_ definitions, respectively. I therefore rename variables accordingly in the `lib/claim` directory.